mirgecom/viscous.py

```diff
@@ -0,0 +1,399 @@
+.. autofunction:: get_local_max_species_diffusivity
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+from grudge.trace_pair import TracePair
+from meshmode.dof_array import thaw, DOFArray
+
+from mirgecom.flux import divergence_flux_central
+from mirgecom.fluid import (
+    velocity_gradient,
+    species_mass_fraction_gradient,
+    make_conserved
+)
+from mirgecom.eos import MixtureEOS
+
+
+def viscous_stress_tensor(discr, eos, cv, grad_cv):
+    r"""Compute the viscous stress tensor.
+
+    The viscous stress tensor $\tau$ is defined by:
+
+    .. math::
+
+        \mathbf{\tau} = \mu\left(\nabla{\mathbf{v}}
+        +\left(\nabla{\mathbf{v}}\right)^T\right) + (\mu_B - \frac{2\mu}{3})
+        \left(\nabla\cdot\mathbf{v}\right)
+
+    Parameters
+    ----------
+    discr: :class:`grudge.eager.EagerDGDiscretization`
+        The discretization to use
+    eos: :class:`~mirgecom.eos.GasEOS`
+        A gas equation of state with a non-empty
+        :class:`~mirgecom.transport.TransportModel`.
+    cv: :class:`~mirgecom.fluid.ConservedVars`
+        Fluid state
+    grad_cv: :class:`~mirgecom.fluid.ConservedVars`
+        Gradient of the fluid state
+
+    Returns
+    -------
+    numpy.ndarray
+        The viscous stress tensor
+    """
+    dim = cv.dim
+    transport = eos.transport_model()
+
+    mu_b = transport.bulk_viscosity(eos, cv)
+    mu = transport.viscosity(eos, cv)
+
+    grad_v = velocity_gradient(discr, cv, grad_cv)
+    div_v = np.trace(grad_v)
+
+    return mu*(grad_v + grad_v.T) + (mu_b - 2*mu/3)*div_v*np.eye(dim)
+
+
+def diffusive_flux(discr, eos, cv, grad_cv):
+    r"""Compute the species diffusive flux vector, ($\mathbf{J}_{\alpha}$).
+
+    The species diffusive flux is defined by:
+
+    .. math::
+
+        \mathbf{J}_{\alpha} = -\rho{d}_{(\alpha)}\nabla{Y_{\alpha}}~~
+        (\mathtt{no~implied~sum}),
```

Comment:
```suggestion
        \mathbf{J}_{\alpha} = -\rho{d}_{(\alpha)}\nabla{Y_{\alpha}},
```
(`\alpha` occurs on both sides of the equation, also wouldn't the same remark be needed lots of other places for consistency? Also `\mathtt` for human-readable text is weird.)

Reply:
Can you explain your reasoning here?  The document states Einstein summation convention is used throughout except where otherwise noted.  This is one of those wheres, and since \alpha is a repeated index here in the term on the right side, the comment is needed afaict.